src/Type/Definition/QueryPlan.php

```diff
@@ -0,0 +1,246 @@
+
+namespace GraphQL\Type\Definition;
+
+use GraphQL\Error\Error;
+use GraphQL\Executor\Values;
+use GraphQL\Language\AST\FieldNode;
+use GraphQL\Language\AST\FragmentDefinitionNode;
+use GraphQL\Language\AST\FragmentSpreadNode;
+use GraphQL\Language\AST\InlineFragmentNode;
+use GraphQL\Language\AST\SelectionSetNode;
+use GraphQL\Type\Schema;
+use function array_filter;
+use function array_key_exists;
+use function array_keys;
+use function array_merge;
+use function array_merge_recursive;
+use function array_unique;
+use function array_values;
+use function count;
+use function in_array;
+use function is_array;
+use function is_numeric;
+
+class QueryPlan
+{
+    /** @var string[][] */
+    private $types = [];
+
+    /** @var Schema */
+    private $schema;
+
+    /** @var mixed[] */
+    private $queryPlan = [];
+
+    /** @var mixed[] */
+    private $variableValues;
+
+    /** @var FragmentDefinitionNode[]|null */
+    private $fragments;
+
+    /**
+     * @param FieldNode[]|null         $fieldNodes
+     * @param mixed[]                  $variableValues
+     * @param FragmentDefinitionNode[] $fragments
+     *
+     * @throws Error
+     */
+    public function __construct(ObjectType $parentType, Schema $schema, $fieldNodes, array $variableValues, array $fragments)
+    {
+        $this->schema         = $schema;
+        $this->variableValues = $variableValues;
+        $this->fragments      = $fragments;
+        $this->analyzeQueryPlan($parentType, $fieldNodes);
+    }
+
+    /**
+     * @return mixed[]
+     */
+    public function queryPlan() : array
+    {
+        return $this->queryPlan;
+    }
+
+    /**
+     * @return string[]
+     */
+    public function referencedTypes() : array
+    {
+        return array_keys($this->types);
+    }
+
+    public function hasType(string $type) : bool
+    {
+        return count(array_filter($this->referencedTypes(), static function ($element) use ($type) {
+                return $type === $element;
+        })) > 0;
+    }
+
+    /**
+     * @return string[]
+     */
+    public function referencedFields() : array
+    {
+        return array_values(array_unique(array_merge(...array_values($this->types))));
+    }
+
+    public function hasField(string $field) : bool
+    {
+        return count(array_filter($this->referencedFields(), static function ($element) use ($field) {
+            return $field === $element;
+        })) > 0;
+    }
+
+    /**
+     * @return string[]
+     */
+    public function subFields(string $typename) : array
+    {
+        if (! array_key_exists($typename, $this->types)) {
+            return [];
+        }
+
+        return $this->types[$typename];
+    }
+
+    /**
+     * @param FieldNode[]|null $fieldNodes
+     *
+     * @throws Error
+     */
+    private function analyzeQueryPlan(ObjectType $parentType, $fieldNodes) : void
+    {
+        $queryPlan = [];
+        /** @var FieldNode $fieldNode */
+        foreach ($fieldNodes as $fieldNode) {
```

Comment:
Please fix typehint or handle nullability, otherwise it's not safe to iterate `$fieldNodes`, not sure what's correct here :)